component/custom/custom.go

```diff
@@ -0,0 +1,96 @@
+   Author : <blobfusedev@microsoft.com>
+
+   Permission is hereby granted, free of charge, to any person obtaining a copy
+   of this software and associated documentation files (the "Software"), to deal
+   in the Software without restriction, including without limitation the rights
+   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+   copies of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in all
+   copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+   SOFTWARE
+*/
+
+package custom
+
+import (
+	"fmt"
+	"os"
+	"plugin"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-storage-fuse/v2/common/log"
+	"github.com/Azure/azure-storage-fuse/v2/exported"
+	"github.com/Azure/azure-storage-fuse/v2/internal"
+)
+
+func initializePlugins() error {
+
+	// Environment variable which expects file path as a colon-separated list of `.so` files.
+	// Example BLOBFUSE_PLUGIN_PATH="/path/to/plugin1.so:/path/to/plugin2.so"
+	pluginFilesPath := os.Getenv("BLOBFUSE_PLUGIN_PATH")
+	if pluginFilesPath == "" {
+		return nil
+	}
+
+	pluginFiles := strings.Split(pluginFilesPath, ":")
+
+	for _, file := range pluginFiles {
+		if strings.HasSuffix(file, ".so") {
+			log.Info("loading plugin %s", file)
+			startTime := time.Now()
+			p, err := plugin.Open(file)
+			if err != nil {
+				return fmt.Errorf("error opening plugin %s: %s", file, err.Error())
+			}
+
+			getExternalComponentFunc, err := p.Lookup("GetExternalComponent")
+			if err != nil {
+				return fmt.Errorf("error looking up GetExternalComponent function in %s: %s", file, err.Error())
+			}
+
+			getExternalComponent, ok := getExternalComponentFunc.(func() (string, func() exported.Component))
+			if !ok {
+				return fmt.Errorf("error casting GetExternalComponent function in %s", file)
+			}
+
+			compName, initExternalComponent := getExternalComponent()
+			internal.AddComponent(compName, initExternalComponent)
+			duration := time.Since(startTime)
+			log.Info("plugin %s loaded successfully in %v", file, duration)
+		} else {
+			return fmt.Errorf("invalid plugin file extension: %s", file)
+		}
+	}
+	return nil
+}
+
+func init() {
+	err := initializePlugins()
+	if err != nil {
+		log.Err("custom::Error initializing plugins: %s", err.Error())
+		fmt.Printf("failed to initialize plugin: %s\n", err.Error())
+		os.Exit(1)
```

Comment:
@vibhansa-msft Is this correct to exit from the blobfuse here?

Reply:
Yes if the customer has chosen to load a component and we fail to do that then we shall exit with throwing an error on console. Looks correct to me. Also, if we do not fail here and allow system to run anyway during init of the pipeline its going to fail as it will not be able to find the component. Better to fail early here.